src/main/java/com/palantir/gradle/versions/lockstate/Line.java

```diff
@@ -46,7 +46,13 @@ default MyModuleIdentifier identifier() {
 
     @Lazy
     default String stringRepresentation() {
-        return String.format(
-                "%s:%s:%s (%s constraints: %s)", group(), name(), version(), numDependents(), dependentsHash());
+        // Three lines of non-changing comments
```

Comment:
Is there somewhere in the default git merge driver's code that makes 3 lines a special amount of context?  I'm wondering where this comes from, and why not 1 or 2 or 4+

Reply:
I'm not super familiar with the `git` source tree, and I haven't been able to find the exact code for default merge driver, but https://github.com/git/git/blob/a9e066fa63149291a55f383cfa113d8bdbdaa6b3/t/t4055-diff-context.sh#L40 is test case `test_expect_success 'the default number of context lines is 3'`

Separately, gnu diff in diffutils also defaults to 3 lines for context per https://git.savannah.gnu.org/cgit/diffutils.git/tree/src/diff.c#n944

Reply:
Not sure if this helps, but since I was pinged, I'll try to answer...

> Is there somewhere in the default git merge driver's code that makes 3 lines a special amount of context? I'm wondering where this comes from, and why not 1 or 2 or 4+

So, the merging of three textual files is done by the xdiff library in Git (which...has a history.  It's an internal copy of libxdiff, which was modified internally.  Other projects have also copied and modified this library.  The upstream project is dead, the code is somewhat inscrutable at times, someone tried to split xdiff back out of Git so it could be shared at https://github.com/libgit2/xdiff but Git still uses their internal copy).  I've rarely touched the xdiff code, so I'm slightly outside my wheelhouse.  However, I have extensively modified the parts of the Git merge machinery that call xdiff, so I know at least a little.

The file of interest is https://git.kernel.org/pub/scm/git/git.git/tree/xdiff/xmerge.c, and you can start with the function xdl_merge().

Digging around a bit, I think the constant of 3 is only relevant in terms of the number of lines of content being printed when creating a diff.  For example, from the Git sources:
```
builtin_diff() ->
  sets xecfg.ctxlen
  xdi_diff_outf() [passing xecfg]->
    xdi_diff() [passing xecfg] ->
      xdl_diff() [passing xefcg] ->
        xdl_do_diff() [NOT passing xefcg] ->
          xdl_recs_cmp()
        emit_func called with xefcg passed
```
whereas for merges:
```
merge_3way() ->
  ll_merge() ->
    xdl_merge() ->
      xdl_do_diff() ->
        xdl_recs_cmp()    
```
but importantly, nowhere in the latter call chain sets up or uses a xdemitconf_t (the type of xecfg), and nothing calls an emit function.  I think the extended conflict range you see likely has more to do with the xdl_change_compact() calls trying to merge nearby groups of changed & unchanged code.

Further a `git grep '\b3\b' xdiff` and investigating all of them, and doing a few other searches, doesn't seem to suggest there's any variable or constant controlling "context" lines, just the number of changed lines (chg[012] is a useful search there).

Basically, I think the behavior of nearby lines being sucked into the conflict region is pretty hard-coded into the algorithm currently in use.

> I'm not super familiar with the git source tree, and I haven't been able to find the exact code for default merge driver, but https://github.com/git/git/blob/a9e066fa63149291a55f383cfa113d8bdbdaa6b3/t/t4055-diff-context.sh#L40 is test case test_expect_success 'the default number of context lines is 3'

Yes, for diff generation, 3 lines of context is the default, but this doesn't affect merge operation.  (It would affect git-apply, but apply is not merge).  Note also that the 3 lines of context default for diffs is documented in the git-config(1) manpage (under `diff.context`) and can be found where diff_context_default is defined near the top of https://git.kernel.org/pub/scm/git/git.git/tree/diff.c.

> By default, Git uses three lines of context when determining merge conflicts. This means that Git compares the changes made in the conflicting files with the version of the file in the common ancestor, as well as the two versions of the file being merged. These three versions of the file, along with the three lines of context around each change, are used to determine how to resolve the conflict. However, it is possible to specify a different number of lines of context using the -U or --unified option when running the git merge command.

In the past, not understanding xdiff, I probably would have thought that everything before the final sentence was correct.  From digging into the sources today, I'm not so sure there is even a "number of context lines" idea involved in the algorithm at all.

The last sentence is simply incorrect; try running `git merge -U8 $BRANCH` or `git merge --unified 8 $BRANCH`. 
 It'll throw an error (`unknown switch 'U'` or `unknown option 'unified'`)